### DISCUSSION TITLE
get peer ID upon handler creation

### DISCUSSION THREAD

**Discussion by drHuangMHT** [libp2p/rust-libp2p#3489] [Q&A] [answered]
`fn new_handler(&mut self)` in trait `NetworkBehaviour` does not provide the peer which the newly created handler is dedicated to. The only way to inform the handler of the peer itself is connected to seems to be emit an event to "request" for peer ID after the handler was created. Is that by design or can be improved potentially by adding another argument to `new_handler()`? When you try to notify a non-exist handler the event will be dropped silently, which will, in my case, cause an oneshot sender be dropped resulting in error on the receiving end, that's not expected.

**Comment by thomaseizinger** [comment, marked as answer]
Currently, this can be achieved by implementing `IntoConnectionHandler` instead.

Once #3254 merges, it will be even easier.

**Reply by drHuangMHT** [reply]
I see. That's fantastic!
